### PR TITLE
Allow specifying run script and copy files phase ordering

### LINKED ISSUE #400: Should allow specifying ordering of copy files phases
From Xcode 10 release notes:

> While a target is performing post-compilation build steps, such as linking and code signing, Xcode’s new build system will begin compiling sources from targets that depend on that target, to improve build speed by using more of your available processor cores. This compilation will not start until any “Run Script” phases from the target have completed. (38772619)

We should allow specifying of the ordering of copy file phases. Scripts will block the compilation of a dependent target, while copy files ~~(probably)~~ doesn't. Being able to determine the order of these can improve build times, but also impact correctness of a build. The copy files phase might have to come before a post-build run script or even before Compile Sources, in order to allow dependent targets to compile correctly.

### PR DESCRIPTION
**Pull request by brentleyjones**
Fixes #400.

- Allows specifying run scripts to run immediately after the Compile Sources phase (which should be used instead of `postbuildScripts` 99% of the time).
- Allows specifying copy files phases to be run before the Compile Sources phase
    - Adjusts modulemap and static library heady copy files phases to be run pre Compile Sources
    - Adjusts the Carthage copy-frameworks to be run sooner

### CODE REVIEW

File: Sources/ProjectSpec/TargetSource.swift [hunk 1]
```diff
@@ -119,6 +133,7 @@ public struct TargetSource: Equatable {
         self.type = type
         self.optional = optional
         self.buildPhase = buildPhase
+        self.headerVisibility = headerVisibility
```
Comment (brentleyjones) [inline Sources/ProjectSpec/TargetSource.swift]: I noticed this wasn't being properly set. Didn't necessarily intend to commit it with this.

File: Sources/XcodeGenKit/SourceGenerator.swift [hunk 2]
```diff
@@ -194,7 +195,8 @@ class SourceGenerator {
                 guard targetType == .staticLibrary else { return nil }
                 return .copyFiles(TargetSource.BuildPhase.CopyFilesSettings(
                     destination: .productsDirectory,
-                    subpath: "include/$(PRODUCT_NAME)"
+                    subpath: "include/$(PRODUCT_NAME)",
+                    phaseOrder: .preCompile
```
Comment (brentleyjones) [inline Sources/XcodeGenKit/SourceGenerator.swift]: This being `.preCompile` is a prime example of why these changes are needed.

**Comment by brentleyjones** [comment]
Even with these changes the new build system is broken. That doesn't make these changes not needed though. Eventually Xcode 10 will be fixed, but being fixed will still warrant these changes.

**Comment by toshi0383** [comment]
@brentleyjones 

> which should be used instead of postbuildScripts 99% of the time

Question: What do you expect for remaining 1% of them?

Also I think `sources`'s `buildPhase` structure is getting too complicated. It should look as same as we see on Xcode.
Ideally I expect a format like this.

```yml
        buildPhase:
          copyFiles:
            destination: productsDirectory
            subpath: include
          compile:      // <- This simply makes the processing order clearer.
          runScript:
            path: my-script.sh
```

**Comment by brentleyjones** [comment]
> Question: What do you expect for remaining 1% of them?

When you want to post-process something that has been copied into the products folder (like adjusting application icons for debug versus releases).

File: Sources/ProjectSpec/TargetSource.swift (hunk 1 above)
Comment (yonaskolb) [reply Sources/ProjectSpec/TargetSource.swift]: I'm ok with it being here

File: Sources/ProjectSpec/Target.swift [hunk 3]
```diff
@@ -32,6 +32,7 @@ public struct Target: ProjectTarget {
     public var directlyEmbedCarthageDependencies: Bool?
     public var requiresObjCLinking: Bool?
     public var prebuildScripts: [BuildScript]
+    public var postCompileScripts: [BuildScript]
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Target.swift]: Now the old `prebuildScripts` and `postbuildScript` names look funny. Maybe we should update them to `preBuildScripts` and `postBuildScript`, and use the old casing as a fallback?
What do you think?

File: Docs/ProjectSpec.md [hunk 4]
```diff
@@ -373,7 +376,7 @@ targets:
 
 ### Build Script
 
-Run script build phases added via **prebuildScripts** or **postBuildScripts**. They run before or after any other build phases respectively and in the order defined. Each script can contain:
+Run script build phases added via **prebuildScripts**, **postCompileScripts**, or **postBuildScripts**. They run before any other build phases, after the Compile Sources build phase, or after any other build phases respectively and in the order defined. Each script can contain:
```
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: Now that we have 3, maybe we need to split them into explicit sections:

Run script build phases can be added at 3 different points in the build:
- **prebuildScripts**: Before any other build phases
- **postCompileScripts** After the compile sources build phase
- **postbuildScripts**: After any other build phases

Each script can contain:

File: Sources/ProjectSpec/Target.swift (hunk 3 above)
Comment (brentleyjones) [reply Sources/ProjectSpec/Target.swift]: Sounds good.

**Comment by brentleyjones** [comment]
@yonaskolb Care to give this another look?

File: Sources/ProjectSpec/Target.swift [hunk 5]
```diff
@@ -279,8 +283,9 @@ extension Target: NamedJSONDictionaryConvertible {
         directlyEmbedCarthageDependencies = jsonDictionary.json(atKeyPath: "directlyEmbedCarthageDependencies")
         requiresObjCLinking = jsonDictionary.json(atKeyPath: "requiresObjCLinking")
 
-        prebuildScripts = jsonDictionary.json(atKeyPath: "prebuildScripts") ?? []
-        postbuildScripts = jsonDictionary.json(atKeyPath: "postbuildScripts") ?? []
+        preBuildScripts = jsonDictionary.json(atKeyPath: "preBuildScripts") ?? jsonDictionary.json(atKeyPath: "prebuildScripts") ?? []
```
Comment (yonaskolb) [inline Sources/ProjectSpec/Target.swift]: Makes me think the `lint` command in https://github.com/yonaskolb/XcodeGen/issues/214 could also be used to highlight deprecations

File: Docs/ProjectSpec.md [hunk 6]
```diff
@@ -373,7 +377,13 @@ targets:
 
 ### Build Script
 
-Run script build phases added via **prebuildScripts** or **postBuildScripts**. They run before or after any other build phases respectively and in the order defined. Each script can contain:
+Run script build phases can be added at 3 different points in the build:
```
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: I guess another option would be to have a single `buildScripts` array with each item having an `order` enum. Is this cleaner and more extensible? Thoughts?

File: Docs/ProjectSpec.md [hunk 7]
```diff
@@ -275,6 +276,9 @@ A source can be provided via a string (the path) or an object of the form:
 			- `sharedSupport`
 			- `plugins`
 		- [ ] **subpath**: **String** - The path inside of the destination to copy the files.
+		- [ ] **phaseOrder**: **String** - When the Copy Files phase should execute in relation to other phases. This can be one of the following values:
```
Comment (yonaskolb) [inline Docs/ProjectSpec.md]: Does this need to be exposed to users? Would it ever be required? I'd rather keep the options we expose low. XcodeGen is supposed to remove all those sorts of decisions.

Comment (brentleyjones) [reply Docs/ProjectSpec.md]: I felt it should be exposed since a user could choose to have a specific set of resources copied before compilation. We currently have those covered, AFAIK, but others might have more exotic needs.

File: Docs/ProjectSpec.md (hunk 6 above)
Comment (brentleyjones) [reply Docs/ProjectSpec.md]: Since we already have the two values, I think this is cleaner. These should be the only 3 spots needed. If we did any refactor I would want to be able to specify the order of all of the phases. It's hard for me to imagine how that would look, and it's probably a big enough breaking change that it would need to be part of a 2.0 release.

File: Docs/ProjectSpec.md (hunk 7 above)
Comment (yonaskolb) [reply Docs/ProjectSpec.md]: I'd rather leave this out for now, and only add it if it someone has a need for it 👍

**Comment by yonaskolb** [comment]
Hi @brentleyjones. Would you be able to get this branch up to date resolve the conflicts?

**Comment by brentleyjones** [comment]
@yonaskolb Yeah, I'll try to get to it today 😄.

**Comment by brentleyjones** [comment]
@yonaskolb Assuming tests pass, I believe this is all merged up. I also removed the ability to specify copy files phase ordering.

File: Docs/ProjectSpec.md (hunk 7 above)
Comment (brentleyjones) [reply Docs/ProjectSpec.md]: @yonaskolb Coming back from WWDC I've learned that the way I'm doing module map copying currently isn't the best. While hacking through a local solution before figuring out how XcodeGen should be updated I ran into wanting to specify the phase order of the copy phase. Essentially I need to copy files before compilation. Wondering how you feel about me exposing this option now?
